src/packs.c

```diff
@@ -60,56 +131,47 @@ static int download_pack(int oldversion, int newversion, char *module, int is_mi
 			return err;
 		}
 		printf("Linked %s to %s\n", url, filename);
+
+		err = finalize_pack_download(module, newversion, filename);
 		free_string(&url);
+		free_string(&filename);
 	} else {
+		struct pack_data *pack_data;
+
 		string_or_die(&url, "%s/%i/pack-%s-from-%i.tar", content_url, newversion, module, oldversion);
 
-		err = swupd_curl_get_file_full(url, filename, NULL, resume_ok);
-		if (err) {
-			if (err == -ENET404) {
-				telemetry(TELEMETRY_WARN, "packmissing",
-					  "url=%s\n", url);
-			}
-			free_string(&url);
-			if ((lstat(filename, &stat) == 0) && (stat.st_size == 0)) {
-				unlink(filename);
-			}
-			free_string(&filename);
-			return err;
+		pack_data = calloc(1, sizeof(struct pack_data));
+		if (!pack_data) {
+			abort();
 		}
-		free_string(&url);
-	}
 
-	fprintf(stderr, "\nExtracting %s pack for version %i\n", module, newversion);
-	err = extract_to(filename, state_dir);
+		pack_data->url = url;
+		pack_data->filename = filename;
+		pack_data->module = module;
+		pack_data->newversion = newversion;
 
-	unlink(filename);
-	/* make a zero sized file to prevent redownload */
-	tarfile = fopen(filename, "w");
-	free_string(&filename);
-	if (tarfile) {
-		fclose(tarfile);
+		err = swupd_curl_parallel_download_enqueue(download_handle, url, filename, NULL, pack_data);
 	}
 
-	// Only negative return values should indicate errors
-	if (err > 0) {
-		return -err;
-	} else {
-		return err;
-	}
+	return err;
 }
 
 /* pull in packs for base and any subscription */
-int download_subscribed_packs(struct list *subs, struct manifest *mom, bool required, bool resume_ok)
+int download_subscribed_packs(struct list *subs, struct manifest *mom, bool required)
 {
 	struct list *iter;
 	struct sub *sub = NULL;
 	int err;
 	int is_mix = 0;
 	unsigned int list_length = list_len(subs);
 	unsigned int complete = 0;
+	void *download_handle;
 
 	fprintf(stderr, "Downloading packs...\n");
+	download_handle = swupd_curl_parallel_download_start(MAX_XFER, MAX_XFER_BOTTOM);
+	download_successful(NULL);
```

Comment:
Why call this now? it will just return false immediately because of the `NULL` param right?

Reply:
My bad. That was a test I forgot to remove. I'll fix this.